fvm/src/machine/manifest.rs

```diff
@@ -0,0 +1,163 @@
+use std::collections::{HashMap, HashSet};
+
+use anyhow::{anyhow, Context};
+use cid::Cid;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::CborStore;
+
+const SINGLETON_ACTOR_NAMES: &[&str] = &[
+    "system",
+    "init",
+    "reward",
+    "cron",
+    "storagepower",
+    "storagemarket",
+    "verifiedregistry",
+];
+
+const ACCOUNT_ACTOR_NAME: &str = "account";
+const INIT_ACTOR_NAME: &str = "init";
+const SYSTEM_ACTOR_NAME: &str = "system";
+
+/// A mapping of builtin actor CIDs to their respective types.
+pub struct Manifest {
+    account_code: Cid,
+    system_code: Cid,
+    init_code: Cid,
+    singletons: HashSet<Cid>,
+
+    by_id: HashMap<u32, Cid>,
+    by_code: HashMap<Cid, u32>,
+}
+
+/// Create an "id CID" (for testing).
+#[cfg(any(feature = "testing", test))]
+const fn id_cid(name: &[u8]) -> Cid {
+    use std::mem;
+
+    use fvm_shared::{IDENTITY_HASH, IPLD_RAW};
+    use multihash::Multihash;
+
+    // This code is ugly because const fns are a bit ugly right now:
+    //
+    // 1. We can't drop the result, we need to forget it.
+    // 2. We can't unwrap, we need to explicitly panic.
+    // 3. Rust can't figure out that "panic" will prevent a drop of the error case.
+    let result = Multihash::wrap(IDENTITY_HASH, name);
+    let k = if let Ok(mh) = &result {
+        Cid::new_v1(IPLD_RAW, *mh)
+    } else {
+        panic!();
+    };
+    mem::forget(result); // This is const, so we don't really care about "leaks".
+
+    k
+}
+
+impl Manifest {
+    #[cfg(any(feature = "testing", test))]
+    pub const DUMMY_CODES: &'static [(&'static str, Cid)] = &[
+        ("system", id_cid(b"fil/test/system")),
+        ("init", id_cid(b"fil/test/init")),
+        ("cron", id_cid(b"fil/test/cron")),
+        ("account", id_cid(b"fil/test/account")),
+    ];
+
+    #[cfg(any(feature = "testing", test))]
+    pub fn dummy() -> Self {
+        Self::new(Self::DUMMY_CODES.iter().copied()).unwrap()
+    }
+
+    /// Load a manifest from the blockstore.
+    pub fn load<B: Blockstore>(bs: &B, root_cid: &Cid, ver: u32) -> anyhow::Result<Manifest> {
+        if ver != 1 {
+            return Err(anyhow!("unsupported manifest version {}", ver));
+        }
+
+        let vec: Vec<(String, Cid)> = match bs.get_cbor(root_cid)? {
+            Some(vec) => vec,
+            None => {
+                return Err(anyhow!("cannot find manifest root cid {}", root_cid));
+            }
+        };
+        Manifest::new(vec)
+    }
+
+    /// Construct a new manifest from actor name/cid tuples.
+    pub fn new(iter: impl IntoIterator<Item = (impl Into<String>, Cid)>) -> anyhow::Result<Self> {
+        let mut by_name = HashMap::new();
+        let mut by_id = HashMap::new();
+        let mut by_code = HashMap::new();
+        for ((name, code_cid), id) in iter.into_iter().zip(1u32..) {
```

Comment:
can we add a comment that this needs to start at 1 or disaster with the 0 is not found semantics?